train.py

```diff
@@ -1,18 +1,389 @@
+
+To run with DDP on 4 gpus on 1 node, example:
+$ torchrun --standalone --nproc_per_node=4 train.py
+
+To run with DDP on 4 gpus across 2 nodes, example:
+- Run on the first (master) node with example IP 123.456.123.456:
+$ torchrun --nproc_per_node=8 --nnodes=2 --node_rank=0 --master_addr=123.456.123.456 --master_port=1234 train.py
+- Run on the worker node:
+$ torchrun --nproc_per_node=8 --nnodes=2 --node_rank=1 --master_addr=123.456.123.456 --master_port=1234 train.py
+(If your cluster does not have Infiniband interconnect prepend NCCL_IB_DISABLE=1)
+"""
+
+import os
+import math
+import time
+from contextlib import nullcontext
+
+import yaml
+import hydra
+import torch
+import wandb
+import numpy as np
 from datasets import load_dataset
+from hydra.utils import to_absolute_path
+from omegaconf import OmegaConf, DictConfig
+from torch.nn.parallel import DistributedDataParallel as DDP
+from torch.distributed import init_process_group, destroy_process_group
+from midi_tokenizers_generation.tokenizer_generator import generate_tokenizer
+
+from model import GPT, GPTConfig
+from data.next_token_dataset import NextTokenDataset
+from tokenized_midi_datasets import OneTimeTokenDataset, TokenizedMidiDataset, ExponentialTimeTokenDataset
+
+tokenizer_name_to_dataset_map: dict[str, TokenizedMidiDataset] = {
+    "NoLossTokenizer": ExponentialTimeTokenDataset,
+    "OneTimeTokenizer": OneTimeTokenDataset,
+}
+
 
+@hydra.main(config_path="configs", config_name="gpt2_noloss_pretraining")
+def main(cfg: DictConfig):
+    # Get the right data for the tokenizer specified in config
+    dataset_builder = tokenizer_name_to_dataset_map[cfg.data.tokenizer]
+    dataset_config = dataset_builder.builder_configs[cfg.data.dataset_name]
+    tokenizer_parameters = dataset_config.tokenizer_parameters
 
-#  prototype - testing dataset building
-def main():
+    # Hydra changes paths - we have to change them back
+    # Load the suitable dataset
+    dataset_path = to_absolute_path(f"./tokenized_midi_datasets/{dataset_builder.__name__}")
     dataset = load_dataset(
-        "tokenized_midi_datsets/OneTimeTokDataset",
-        name="debugging",
-        trust_remote_code=True,
+        dataset_path,
+        name=cfg.data.dataset_name,
         num_proc=8,
+        trust_remote_code=True,
+    )
+    total_tokens = dataset_config.sequence_length * dataset["train"].num_rows
+    print(f"tokens in a training dataset: {total_tokens}")
+    if cfg.architecture != "custom":
+        architecture_config_path = f"configs/architectures/{cfg.architecture}.yaml"
+        architecture_config_path = to_absolute_path(architecture_config_path)
+        with open(architecture_config_path, "r") as file:
+            architecture_config = yaml.safe_load(file)
+
+        # override model config
+        cfg.model.n_layer = architecture_config["n_layer"]
+        cfg.model.n_head = architecture_config["n_head"]
+        cfg.model.n_embd = architecture_config["n_embd"]
```

Comment:
Look like this is not using hydra correctly - you should be able to load any architecture config without manually reading it as yaml, but by defining a proper relation in hydra: https://hydra.cc/docs/patterns/configuring_experiments/